base_layer/core/src/proof_of_work/target_difficulty.rs

```diff
@@ -20,28 +20,92 @@
 // WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 // USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-use crate::{
-    consensus::ConsensusManagerError,
-    proof_of_work::{difficulty::DifficultyAdjustment, lwma_diff::LinearWeightedMovingAverage, Difficulty},
-};
+use crate::proof_of_work::{difficulty::DifficultyAdjustment, lwma_diff::LinearWeightedMovingAverage, Difficulty};
 use std::cmp;
 use tari_crypto::tari_utilities::epoch_time::EpochTime;
 
-/// Returns the estimated target difficulty for the provided set of target difficulties.
-pub fn get_target_difficulty(
-    target_difficulties: Vec<(EpochTime, Difficulty)>,
-    block_window: usize,
-    target_time: u64,
+#[derive(Debug, Clone)]
+pub struct TargetDifficulty {
+    lwma: LinearWeightedMovingAverage,
     min_difficulty: Difficulty,
     max_difficulty: Difficulty,
-    max_block_time: u64,
-) -> Result<Difficulty, ConsensusManagerError>
-{
-    let mut lwma = LinearWeightedMovingAverage::new(block_window, target_time, min_difficulty, max_block_time);
-    for (epoch, difficulty) in target_difficulties {
-        lwma.add(epoch, difficulty)?;
-    }
-    let mut target_difficulty = lwma.get_difficulty();
-    target_difficulty = cmp::min(max_difficulty, cmp::max(target_difficulty, min_difficulty));
-    Ok(target_difficulty)
+}
+
+impl TargetDifficulty {
```

Comment:
Perhaps this describes the window rather than a single TargetDifficulty value 
```suggestion
impl TargetDifficultyWindow {
```

Reply:
Yeah not opposed to that name